app/Actions/GrantRayTrialLicenseAction.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace App\Actions;
+
+use App\Domain\Shop\Actions\CreateLicenseAction;
+use App\Domain\Shop\Models\License;
+use App\Domain\Shop\Models\Purchasable;
+use App\Domain\Shop\Models\Purchase;
+use App\Domain\Shop\Models\PurchaseAssignment;
+use App\Mail\RayTrialLicenseGrantedMail;
+use App\Models\User;
+use Illuminate\Contracts\Database\Eloquent\Builder;
+use Illuminate\Support\Facades\Mail;
+
+class GrantRayTrialLicenseAction
+{
+    public function execute(User $user): ?License
+    {
+        if ($this->alreadyTriedRay($user)) {
+            return null;
+        }
+
+        $rayPurchasable = Purchasable::query()->where('title', 'Ray license')->first();
+
+        $purchase = Purchase::create([
+            'user_id' => $user->id,
+            'purchasable_id' => $rayPurchasable->id,
+            'paddle_fee' => 0,
+            'earnings' => 0,
+        ]);
+
+        $assignment = PurchaseAssignment::create([
+            'purchase_id' => $purchase->id,
+            'purchasable_id' => $rayPurchasable->id,
+            'user_id' => $user->id,
+        ]);
+
+        Mail::to($user->email)->send(new RayTrialLicenseGrantedMail());
+
+        return app(CreateLicenseAction::class)->execute($assignment, now()->addMonth());
+    }
+
+    protected function alreadyTriedRay(User $user): bool
+    {
+        $rayPurchaseableIds = Purchasable::query()
+            ->whereHas('product', function (Builder $query) {
+                $query->where('title', 'Ray');
+            })
+            ->pluck('id');
+
+        return $user
+                ->assignments()
+                ->whereIn('purchasable_id', $rayPurchaseableIds)
+                ->count() > 0;
```

Comment:
This could be done with `->exists()` as well which should also be a bit faster

Reply:
Good one!